core/rmutex.c

```diff
@@ -0,0 +1,165 @@
+ *
+ * The recursive mutex implementation is inspired by the implementetaion of
+ * Nick v. IJzendoorn <nijzendoorn@engineering-spirit.nl>
+ * @see https://github.com/RIOT-OS/RIOT/pull/4529/files#diff-8f48e1b9ed7a0a48d0c686a87cc5084eR35
+ *
+ */
+
+#include <stdio.h>
+#include <inttypes.h>
+
+#include "rmutex.h"
+#include "thread.h"
+#include "assert.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+void rmutex_lock(rmutex_t *rmutex)
+{
+    kernel_pid_t owner;
+
+    /* try to lock the mutex */
+    DEBUG("rmutex %" PRIi16" : trylock\n", thread_getpid());
+    if (mutex_trylock(&rmutex->mutex) == 0) {
+        DEBUG("rmutex %" PRIi16" : mutex already held\n", thread_getpid());
+        /* Mutex is already held
+         *
+         * Case 1: Mutex is not held by me
+         *     Condition 1: holds
+         *     rmutex->owner != thread_getpid()
+         *
+         * Note for Case 1:
+         *
+         *     As a consequence it is necessaray to call
+         *     mutex_lock(). However the read access to owner is not
+         *     locked, and owner can be changed by a thread that is
+         *     holding the lock (e.g.: holder unlocks the mutex, new
+         *     holder aquired the lock). The atomic access strategy
+         *     'relaxed' ensures, that the value of rmutex->owner is read
+         *     consistent.
+         *
+         *     It is not necessary to synchronize (make written values
+         *     visible) read/write with other threads, because every
+         *     write by other threads let evaluate Condition 1 to
+         *     false. They all write either KERNEL_PID_UNDEF or the
+         *     pid of the other thread.
+         *
+         *     Other threads never set rmutex->owner to the pid of the
+         *     current thread. Hence, it is guaranteed that mutex_lock
+         *     is eventually called.
+         *
+         * Case 2: Mutex is held be me (relock)
+         *     Condition 2: holds
+         *     rmutex->owner == thread_getpid()
+         *
+         * Note for Case 1:
+         *
+         *     Because the mutex rmutex->owner is only written be the
+         *     owner (me), rmutex->owner stays constant througout the
+         *     complete call and rmutex->refcount is protected
+         *     (read/write) by the mutex.
+         */
+
+        /* ensure that owner is read atomically, since I need a consistent value */
+        owner = atomic_load_explicit( &rmutex->owner, memory_order_relaxed);
+        DEBUG("rmutex %" PRIi16" : mutex held by %" PRIi16" \n", thread_getpid(), owner);
+
+        /* Case 1: Mutex is not held by me */
+        if ( owner != thread_getpid() ) {
+            /* wait for the mutex */
+            DEBUG("rmutex %" PRIi16" : locking mutex\n", thread_getpid());
+
+            mutex_lock(&rmutex->mutex);
+        }
+        /* Case 2: Mutex is held be me (relock) */
+        else {
+            assert(rmutex->refcount>0);
```

Comment:
spaces before and after ```>```. (as it's an unsigned integer, I'd be happy with removing the comparison altogether.)